### DISCUSSION TITLE
selectAtom  can write it?

### DISCUSSION THREAD

**Discussion by iambinlin** [pmndrs/jotai#2148] [Q&A] [answered]
Hello, I am migrating from `recoil` to `jotai`, can I ask if `selectAtom` has write capability, or please advise me how to make it have write capability.

**In Recoil**

```ts
const tempFahrenheit = atom({
  key: 'tempFahrenheit',
  default: 32,
});

const tempCelsius = selector({
  key: 'tempCelsius',
  get: ({get}) => ((get(tempFahrenheit) - 32) * 5) / 9,
  set: ({set}, newValue) =>
    set(
      tempFahrenheit,
      newValue instanceof DefaultValue ? newValue : (newValue * 9) / 5 + 32
    ),
});

```

**In Jotai**

```ts
import { selectAtom } from 'jotai/utils'
import { atom } from 'jotai'


export const baseAtom = atom(Math.random());
export const baseAtom1 = atom(Math.random());

export const oneAtom = selectAtom(baseAtom, val => val);

export const twoAtom = selectAtom(atom(get => ({
    a: get(oneAtom)*3,
    b: get(baseAtom1)*2
}), (get, set, payload) => {
    console.log(payload, 'lplplp')
    set(baseAtom1, 99999)
}), val => val.a + val.b);
```
<img width="1152" alt="image" src="https://github.com/pmndrs/jotai/assets/39019913/d0d017f0-6611-430e-9af2-d97cd3589a25">

Online Demo: [https://symmetrical-space-sniffle-9q4rrpp479whrr4.github.dev/](https://symmetrical-space-sniffle-9q4rrpp479whrr4.github.dev/)

But，I received this error in the console.
I understand that the atom I entered is indeed readable and writable, is my usage wrong? I would be very grateful if you could answer if you have time.

**Comment by dai-shi** [comment, marked as answer]
Sorry for the confusion. Please don't use `selectAtom` for such use cases. It's a hacky util to accept `equaltiyFn`. We don't usually need it. It's my regret to name it like that.

Just use `atom((get) => ..., (get, set, arg) => ...)`.

(So, it means there's room for doc improvement.)

**Reply by iambinlin** [reply]
It's ok, Thank you for replying. As the transition to migrate the Recoil, I have already converted 80% of cases very smoothly, and now I'm trying to convert unusual cases. Thank you for your great contribution to this project.

**Reply by iambinlin** [reply]
~~I understand that strictly speaking, for another API: atomFamily, only active reading ability and passive writing ability can be achieved. For active writing ability, it may still need to be implemented by atom itself. I don't use it very deeply. If there is any misunderstanding, you can correct me.~~

<img width="775" alt="image" src="https://github.com/pmndrs/jotai/assets/39019913/ed04a859-2eda-4611-840b-68d4b447e018">

~~A more accurate description would be that reading atomic behavior is similar to `Recoil`.~~

**Reply by dai-shi** [reply]
atomFamily in Jotai is different from Recoil's one, AFAIU. Maybe we should remove that description.

**Comment by iambinlin** [comment]
> for another API: atomFamily, only active reading ability and passive writing ability can be achieved. For active writing ability, it may still need to be implemented by atom itself. 

Sorry, I misunderstood here. In fact, both `jotai` and `atomFamily` have the same active read-write ability. There are problems with my usage and understanding.

I seem to understand some that you said *atomFamily in Jotai is different from Recoil's one*, thank you for your guidance.

-----

In addition，I took a look at the `atomFamily` of `recoil`.
from :[source code](https://github.com/facebookexperimental/Recoil/blob/main/packages/recoil/caches/Recoil_MapCache.js)
```ts
class MapCache<K, V> {
  _map: Map<mixed, V>;
  _keyMapper: K => mixed;
  constructor(options?: Options<K>) {
    this._map = new Map<mixed, V>();
    this._keyMapper = options?.mapKey ?? (v => v);
  }
  get(key: K): ?V {
    return this._map.get(this._keyMapper(key));
  }

  set(key: K, val: V): void {
    this._map.set(this._keyMapper(key), val);
  }
}
```

`_keyMapper` is for filtering and formatting stable `Map key`, while `areEqual` cares more about comparing `params` to get the `Map value`, if I understand it wrong, I hope you can criticize and correct me.

The function of `areEqual` actually has the same capability as `_keyMapper`, and I completely agree with the description in the document. Sorry for my misunderstanding.😭
